pyg_lib/csrc/sampler/cpu/neighbor_kernel.cpp

```diff
@@ -1,21 +1,147 @@
+            if (directed) {
+              rows.push_back(i);
+              cols.push_back(res.first->second);
+              edges.push_back(offset);
+            }
+          }
+        } else if (replace) {
+          for (int64_t j = 0; j < num_samples; j++) {
+            const int64_t offset = row_start + uniform_randint(row_count);
+            const int64_t& v = col_data[offset];
+            const auto res = to_local_node.insert({v, samples.size()});
+            if (res.second)
+              samples.push_back(v);
+            if (directed) {
+              cols.push_back(i);
+              rows.push_back(res.first->second);
+              edges.push_back(offset);
+            }
+          }
+        } else {
+          std::unordered_set<int64_t> rnd_indices;
+          for (int64_t j = row_count - num_samples; j < row_count; j++) {
+            int64_t rnd = uniform_randint(j);
+            if (!rnd_indices.insert(rnd).second) {
+              rnd = j;
+              rnd_indices.insert(j);
+            }
+            const int64_t offset = row_start + rnd;
+            const int64_t& v = col_data[offset];
+            const auto res = to_local_node.insert({v, samples.size()});
+            if (res.second)
+              samples.push_back(v);
+            if (directed) {
+              rows.push_back(i);
+              cols.push_back(res.first->second);
+              edges.push_back(offset);
+            }
+          }
+        }
+      }
+      begin = end, end = samples.size();
+    }
+
+    if (!directed) {
+      phmap::flat_hash_map<int64_t, int64_t>::iterator iter;
+      for (int64_t i = 0; i < static_cast<int64_t>(samples.size()); i++) {
+        const auto& w = samples[i];
+        const auto& row_start = rowptr_data[w];
+        const auto& row_end = rowptr_data[w + 1];
+        for (int64_t offset = row_start; offset < row_end; offset++) {
+          const auto& v = col_data[offset];
+          iter = to_local_node.find(v);
+          if (iter != to_local_node.end()) {
+            cols.push_back(iter->second);
+            rows.push_back(i);
+            edges.push_back(offset);
+          }
+        }
+      }
+    }
+  });
+  return std::make_tuple(from_vector<int64_t>(rows), from_vector<int64_t>(cols),
+                         from_vector<int64_t>(samples),
+                         from_vector<int64_t>(edges));
+}
+
 std::tuple<at::Tensor, at::Tensor, at::Tensor, c10::optional<at::Tensor>>
 neighbor_sample_kernel(const at::Tensor& rowptr,
                        const at::Tensor& col,
                        const at::Tensor& seed,
-                       const std::vector<int64_t> num_neighbors,
+                       const std::vector<int64_t>& num_neighbors,
                        bool replace,
                        bool directed,
                        bool isolated,
                        bool return_edge_id) {
-  return std::make_tuple(rowptr, col, seed, at::nullopt);
+  if (replace && directed) {
```

Comment:
Raise an error in case `isolated` is set.

Reply:
At the moment it always should raise an error?

Reply:
Can you clarify? The argument is ignored currently, while `sample` implements the `isolated=False` case.

Reply:
Ok, I have added an error. Please let me know if this is what You have on mind.